### PR TITLE
crystal: init at 0.29.0

### LINKED ISSUE #63039: Crystal: Missing YAML dependency
[Crystal] requires [LibYAML] for its [YAML][Crystal/YAML] module.

[Crystal]: https://crystal-lang.org
[Crystal/YAML]: https://crystal-lang.org/api/YAML.html
[LibYAML]: https://pyyaml.org/wiki/LibYAML

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

Supersedes #63427
Fixes #63039

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by jonringer** [comment]
This seems to break the scry package
```
builder for '/nix/store/lidci1fs0zfb6h8ypykv1q62bfysx175-scry-0.7.1.20180919.drv' failed with exit code 1; last 10 log lines:

  require "./config"
  ^

  in /nix/store/p9zj24ygr8xr633a6xjrlb7k30gxm7ab-crystal-0.29.0/lib/crystal/compiler/crystal/config.cr:37: undefined constant Crystal::Codegen::Target

      @@default_target : Crystal::Codegen::Target?
                         ^~~~~~~~~~~~~~~~~~~~~~~~
cannot build derivation '/nix/store/1p3z9zp6my9r15d2ny5ml165a74nvmyv-env.drv': 1 dependencies couldn't be built
[10 built (1 failed), 38 copied (1233.4 MiB), 213.1 MiB DL]
error: build of '/nix/store/1p3z9zp6my9r15d2ny5ml165a74nvmyv-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/63725
1 package failed to build:
scry
```

**Comment by jonringer** [comment]
I would try updating the scry package to `0.8`, see if that fixes anything.

**Comment by peterhoeg** [comment]
> I would try updating the scry package to `0.8`, see if that fixes anything.

It doesn't - sorry, I had a few different commits here as I am using it on both stable and unstable and not all of it made it over... I have a fix for scry. Coming shortly.

File: pkgs/development/compilers/crystal/default.nix [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
-, gmp, openssl, readline, tzdata, libxml2, libyaml
+, coreutils, git, gmp, nettools, openssl, readline, tzdata, libxml2, libyaml
 , boehmgc, libatomic_ops, pcre, libevent, libiconv, llvm, clang, which, zlib }:
 
 # We need multiple binaries as a given binary isn't always able to build
-# (even slightly) older or newer version.
+# (even slightly) older or newer versions.
 # - 0.26.1 can build 0.25.x and 0.26.x but not 0.27.x
-# - 0.27.2 can build 0.27.x but not 0.25.x and 0.26.x
+# - 0.27.2 can build 0.27.x but not 0.25.x, 0.26.x and 0.29.x
```
Comment (bcardiff) [inline pkgs/development/compilers/crystal/default.nix]: I'm intrigued by how crystal is starting to become maintained in nixpkgs.

Is there are stable change to be done once a new version is released?

Usually the last `0.m.p` version should be used to build any `0.(m+1).q` version. That is how Crystal releases works. That is the rule that should be followed since the compiler is bootstrapped. Although in the past there were some hiccups with this rule.

In brew, when `0.m.p` is released we built that tag using the GitHub binary release of the latest `0.m-1.q` version.
But from what I "understand" from the formula after a first version of the compiler got into nix the new ones are built after it. Having a new lineage. Am I right?

**Comment by peterhoeg** [comment]
Answering out of order, as that makes things a little easier to follow (for me at least).

>   But from what I "understand" from the formula after a first version of the
>   compiler got into nix the new ones are built after it. Having a new
>   lineage. Am I right?

No - the previous versions of "our" compiler are not used for building crystal - we use the binary distribution. If the latest binary we have is not able to, we will simply bump the binary to the latest and then use that.

>   That is how Crystal releases works. That is the rule that should
>   be followed since the compiler is bootstrapped.

Is it an issue that we don't do it that way?

>   Is there are stable change to be done once a new version is released?

It's always a balancing act between the number of packages we are interested in carrying and what the needs of the users are.

Basically, how I do it is this way:

- when stable is branched off from unstable (near a release of NixOS), this is the version that will stay for the lifetime of stable (6 months until the next stable).
- in unstable we have at minimum the latest version plus whatever version was in the last stable.

We don't backport new crystal versions to stable. Although it's trivial to do, backports are reserved for security fixes.

**Comment by bcardiff** [comment]
Thanks for the clarifications and porting crystal to nixpkgs.

> >   That is how Crystal releases works. That is the rule that should
> >   be followed since the compiler is bootstrapped.
>
> Is it an issue that we don't do it that way?

It's not necessarily an issue. From the comments, it sounded to me like there was some research on which compiler should be used. It seems like a trial an error procedure is made to check if a bump is needed.

As long as the compiler specs passes and the std passes with the new compiler then the behavior should be good enough. Ideally, the progress of the compiler should be guaranteed (ie: the new compiler can build itself again).

**Comment by peterhoeg** [comment]
> It seems like a trial an error procedure is made to check if a bump is needed.

That's exactly how it was done, so your explanation clears that up. I'll update the comment in there.

> As long as the compiler specs passes and the std passes with the new compiler then the behavior should be good enough.

Yeah, that's the problem. It doesn't. We're pretty close but not quite there. It basically has to do path assumptions in the specs that fail when we build using nix.


 Ideally, the progress of the compiler should be guaranteed (ie: the new compiler can build itself again).
